src/internal/connector/onedrive/url_cache.go

```diff
@@ -0,0 +1,275 @@
+	deltaQueryCount int
+
+	svc           graph.Servicer
+	itemPagerFunc func(
+		servicer graph.Servicer,
+		driveID, link string,
+	) itemPager
+
+	errors *fault.Bus
+}
+
+// newURLache creates a new URL cache for the specified drive ID
+func newURLCache(
+	driveID string,
+	refreshInterval time.Duration,
+	svc graph.Servicer,
+	itemPagerFunc func(
+		servicer graph.Servicer,
+		driveID, link string,
+	) itemPager,
+) (*urlCache, error) {
+	err := validateCacheParams(
+		driveID,
+		refreshInterval,
+		svc,
+		itemPagerFunc)
+	if err != nil {
+		return nil, clues.Wrap(err, "invalid cache parameters")
+	}
+
+	return &urlCache{
+			urlMap:          make(map[string]itemProperties),
+			lastRefreshTime: time.Time{},
+			driveID:         driveID,
+			refreshInterval: refreshInterval,
+			svc:             svc,
+			itemPagerFunc:   itemPagerFunc,
+			errors:          fault.New(false),
+		},
+		nil
+}
+
+// validateCacheParams validates the parameters passed to newURLCache
+func validateCacheParams(
+	driveID string,
+	refreshInterval time.Duration,
+	svc graph.Servicer,
+	itemPagerFunc func(
+		servicer graph.Servicer,
+		driveID, link string,
+	) itemPager,
+) error {
+	if len(driveID) == 0 {
+		return clues.New("drive id is empty")
+	}
+
+	if refreshInterval < 0 {
+		return clues.New("invalid refresh interval")
+	}
+
+	if svc == nil {
+		return clues.New("nil graph servicer")
+	}
+
+	if itemPagerFunc == nil {
+		return clues.New("nil item pager")
+	}
+
+	return nil
+}
+
+// getItemProps returns the item properties for the specified drive item ID
+func (uc *urlCache) getItemProperties(
+	ctx context.Context,
+	itemID string,
+) (*itemProperties, error) {
+	if len(itemID) == 0 {
+		return nil, clues.New("item id is empty")
+	}
+
+	ctx = clues.Add(ctx, "drive_id", uc.driveID)
+
+	// Lazy refresh
+	if uc.needsRefresh() {
+		err := uc.refreshCache(ctx)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	url, err := uc.readCache(ctx, itemID)
+	if err != nil {
+		return nil, err
+	}
+
+	return url, nil
+}
+
+// needsRefresh returns true if the cache is empty or if refresh interval has
+// elapsed
+func (uc *urlCache) needsRefresh() bool {
+	uc.cacheLock.RLock()
+	defer uc.cacheLock.RUnlock()
+
+	return len(uc.urlMap) == 0 || time.Since(uc.lastRefreshTime) > uc.refreshInterval
+}
+
+// refreshCache refreshes the URL cache by performing a delta query.
+func (uc *urlCache) refreshCache(
+	ctx context.Context,
+) error {
+	// Acquire mutex to prevent multiple threads from refreshing the
+	// cache at the same time
+	uc.refreshMutex.Lock()
+	defer uc.refreshMutex.Unlock()
+
+	// If the cache was refreshed by another thread while we were waiting
+	// to acquire mutex, return
+	if !uc.needsRefresh() {
+		return nil
+	}
+
+	// Hold cache lock in write mode for the entire duration of the refresh.
+	// This is to prevent other threads from reading the cache while it is
+	// being updated page by page
+	uc.cacheLock.Lock()
+	defer uc.cacheLock.Unlock()
```

Comment:
Based on this implementation, I think you could reduce the two mutexes to a single [RWMutex](https://pkg.go.dev/sync#RWMutex).

The lock on 158 could go away.  146 would use the write lock.  134 and 210 would use the read lock (though you might need to have line 151 call the needsRefresh conditional directly, instead of through the locked needsRefresh.  Or, alternatively, move the needsRefresh read lock to 115).

This reduces the implementation to a single mutex, which will help avoid deadlock and minimize maintenance confusion.

Reply:
Great call out! Those two locks are added by design. I originally had planned to use a single rwMutex lock, but it complicates code quite a bit.
For e.g. consider that `needsRefresh` is lockless, and caller can hold either RLock() or Lock() while calling it. This is needed because we can't acquire RLock() in `needsRefresh` if we are holding Lock().

Now consider line 116 onwards code, copied below with comments. 

```
       // hold RLock()
	if uc.needsRefresh() {
                // RUnlock()
                
		err := uc.refreshCache(ctx) // Assume refreshCache holds Lock() throughout
		if err != nil {
			return nil, err
		}
	} else {
           //RUnlock()
        }
        
	url, err := uc.readCache(ctx, itemID) // Assume readCache holds RLock() throughout.
```

Also, I don't think there is a deadlock possibility here, since we always acquire these 2 locks(refreshMutex, and cacheLock.Lock()) in the same order. Do correct me if I am wrong.

Reply:
I put in some more thought for switching to a single rwMutex. I think it's doable without compromising code readability and maintainability. 
I'll add that in a new commit.

Reply:
if we're thinking about having more complex locking where some functions may require the mutex to already be held it may be worth adding the [checklocks linter](https://pkg.go.dev/gvisor.dev/gvisor/tools/checklocks#section-readme). It requires annotations in code comments, but it can help make sure we don't accidentally forget to grab a lock when we should have

Unfortunately it's not integrated with golangci-lint, but it seems fairly easy to download/run. If we add this it should be done in a separate PR

Reply:
Thanks @ashmrtn! I'll look into this.

Reply:
Update: I'm keeping two locks for now to finish this PR faster. IMO it's not going to cause deadlocks, or impact cache perf. 

I'll iterate on it in a later PR and move to a single rw lock, along with Ashlie's suggested checklocks lint.

Reply:
Going to approve: should be good to move ahead with the current implementation and come back to this.

If you would be so kind, can you please create a tech debt ticket to investigate the double-mutex/single-mutex approach?